### ISSUE TITLE
`libupb` is hard bound in cmake builds

### ISSUE THREAD

**Issue by HofiOne** [protocolbuffers/protobuf#18307]
**What version of protobuf and what language are you using?**
Version: main
Language: C++

**What operating system (Linux, Windows, ...) and version?**
macOS Sonoma 14.6.1 (23G93) - Darwin Kernel Version 23.6.0: Mon Jul 29 21:14:30 PDT 2024; root:xnu-10063.141.2~1/RELEASE_ARM64_T6000 arm64

FreeBSD 14.1-RELEASE releng/14.1-n267679-10e31f0946d8 GENERIC amd64

**What runtime / compiler are you using (e.g., python version or gcc version)**
macOS: Apple clang version 15.0.0 (clang-1500.3.9.4)
FreeBSD: FreeBSD clang version 18.1.5

**What did you do?**
Trying to use protobuf (latest HomeBrew and BSD ports version) for building syslog-ng on macOS and FreeBSD using clang with grpc support.

Also built protobuf from source (main branch), the same result.

1. Try to build syslog-ng (https://github.com/syslog-ng/syslog-ng) master branch using cmake with grpc support
2. During the configure phase one will see the following error
```
CMake Error at /usr/local/lib/cmake/protobuf/protobuf-targets.cmake:42 (message):
  Some (but not all) targets in this export set were already defined.

  Targets Defined: protobuf::libprotobuf-lite, protobuf::libprotobuf,
  protobuf::libprotoc, protobuf::protoc

  Targets not yet defined: protobuf::libupb, protobuf::protoc-gen-upb,
  protobuf::protoc-gen-upbdefs, protobuf::protoc-gen-upb_minitable
```

**Anything else we should know about your project / environment**

The error is caused by the unnecessarily forced `libupb` targets in the `${INSTALL_TARGET}/lib/cmake/protobuf/protobuf-targets.cmake` file.

Some projects, like syslog-ng might not use `libupb` at all.

Removing any references to the 
`protobuf::libupb, protobuf::protoc-gen-upb, protobuf::protoc-gen-upbdefs, protobuf::protoc-gen-upb_minitable` targets will solve the issue, syslog-ng can be built perfectly with grpc support

Please make `libupb` usage optional (probably a build option like `-Dprotobuf_USE_LIBUPB=OFF` could do the trick).

**Comment by HofiOne** [comment]
> We triage inactive PRs and issues in order to make it easier to find active work. If this issue should remain active or becomes active again, please add a comment.
> 
> This issue is labeled `inactive` because the last activity was over 90 days ago. This issue will be closed and archived after 14 additional days without activity.

it would be nice to have this fixed

**Comment by tchernobog** [comment]
I assume this line is the culprit: https://github.com/protocolbuffers/protobuf/blob/d054e536529e266744bf155e2b9253cc9f6b4220/cmake/install.cmake#L41

If the comment at https://github.com/protocolbuffers/protobuf/blob/d054e536529e266744bf155e2b9253cc9f6b4220/cmake/libupb.cmake#L15 is correct, then that line should be removed (it is a private dependency only).

Else I guess it needs to be installed and exported along with the other targets.

Which one is the desirable behavior?

**Comment by HofiOne** [comment]
> I assume this line is the culprit:
> 
> [protobuf/cmake/install.cmake](https://github.com/protocolbuffers/protobuf/blob/d054e536529e266744bf155e2b9253cc9f6b4220/cmake/install.cmake#L41)
> 
> Line 41 in [d054e53](/protocolbuffers/protobuf/commit/d054e536529e266744bf155e2b9253cc9f6b4220)
> 	   list(APPEND _protobuf_libraries libupb) 
> 

the whole section looks like this

if (protobuf_BUILD_LIBUPB)
  list(APPEND _protobuf_libraries libupb)
endif ()

so I guess it takes effect only if `protobuf_BUILD_LIBUPB` is defined that, I guess again, comes from some command line options at build time which just simply has to be turned off if not needed

**Comment by tchernobog** [comment]
> > I assume this line is the culprit:
> > [protobuf/cmake/install.cmake](https://github.com/protocolbuffers/protobuf/blob/d054e536529e266744bf155e2b9253cc9f6b4220/cmake/install.cmake#L41)
> > Line 41 in [d054e53](/protocolbuffers/protobuf/commit/d054e536529e266744bf155e2b9253cc9f6b4220)
> > list(APPEND _protobuf_libraries libupb)
> 
> the whole section looks like this
> 
> if (protobuf_BUILD_LIBUPB) list(APPEND _protobuf_libraries libupb) endif ()
> 
> so I guess it takes effect only if `protobuf_BUILD_LIBUPB` is defined that, I guess again, comes from some command line options at build time which just simply has to be turned off if not needed

No, that behavior is still wrong because you are stating that the upb library, if built, is installed and available to all interface users, and this is not the case.

**Comment by HofiOne** [comment]
The removal should do the trick. If someone needs LIBUPB as well they can add it explicitly in the given project manually.

**Comment by tchernobog** [comment]
I added a patch for convenience, but since I am at the moment not willing to sign a CLA, feel free to port the trivial patch over.

**Comment by bruno-at-bareos** [comment]
Any news on this ?

**Comment by bruno-at-bareos** [comment]
Unfortunately still present in 1.70.1 release.

**Comment by strophy** [comment]
I have encountered this error while attempting to bump protobuf from 26.1 to 27.5 in Alpine Linux. As part of this work, dependent packages need to be rebuilt, and https://github.com/microsoft/onnxruntime fails with:

```
-- Loading Dependencies URLs ...
-- Loading Dependencies ...
-- Found Protobuf: /usr/lib/libprotobuf.so (found version "5.27.5")
-- Found ICU: /usr/include (found version "74.2") found components: uc
-- Found nlohmann_json: /usr/share/cmake/nlohmann_json/nlohmann_jsonConfig.cmake (found version "3.11.3")
-- Found GTest: /usr/lib/cmake/GTest/GTestConfig.cmake (found version "1.15.2")
-- Found ZLIB: /usr/lib/libz.so (found version "1.3.1")
CMake Error at /usr/lib/cmake/protobuf/protobuf-targets.cmake:42 (message):
  Some (but not all) targets in this export set were already defined.

  Targets Defined: protobuf::libprotobuf-lite, protobuf::libprotobuf,
  protobuf::libprotoc, protobuf::protoc

  Targets not yet defined: protobuf::libupb, protobuf::protoc-gen-upb,
  protobuf::protoc-gen-upbdefs, protobuf::protoc-gen-upb_minitable

Call Stack (most recent call first):
  /usr/lib/cmake/protobuf/protobuf-config.cmake:16 (include)
  /usr/share/cmake/Modules/FetchContent.cmake:2472 (find_package)
  /usr/share/cmake/Modules/FetchContent.cmake:2362 (__FetchContent_MakeAvailable_find_package)
  external/helper_functions.cmake:22 (FetchContent_MakeAvailable)
  external/onnxruntime_external_deps.cmake:231 (onnxruntime_fetchcontent_makeavailable)
  CMakeLists.txt:694 (include)
```

I'm not clear from the above discussion if the fix needs to be applied in protobuf or onnxruntime. I have tried [patching](https://github.com/tchernobog/protobuf/commit/c7addd6ba3790f0dd0381e1011537880a36febb5) protobuf 27 before including it as a dependency for onnxruntime without success, and setting `-Dprotobuf_BUILD_LIBUPB=OFF` when building onnxruntime also doesn't work.

**Comment by strophy** [comment]
@haberman my repro above involves onnxruntime, not grpc

**Comment by czln** [comment]
my workaround:
use `find_package(Protobuf REQUIRED)` instead of `find_package(Protobuf CONFIG REQUIRED)`
